### PR TITLE
Blazor FocusAsync coverage

### LINKED ISSUE #20195: ElementReference Validity
Is the following accurate, or is it also true that `ElementReference`s are valid any time after render?

> An ElementReference is only guaranteed valid in a component's OnAfterRender method 

I'm currently using an `ElementReference` in a function that's called when the user clicks something and haven't had any problems.

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: aada0049-99a8-2f91-5d60-6970474e2371
* Version Independent ID: c29fb1a8-028f-6efd-f1bc-e889fb362311
* Content: [Call JavaScript functions from .NET methods in ASP.NET Core Blazor](https://docs.microsoft.com/en-us/aspnet/core/blazor/call-javascript-from-dotnet?view=aspnetcore-5.0)
* Content Source: [aspnetcore/blazor/call-javascript-from-dotnet.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/call-javascript-from-dotnet.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### LINKED ISSUE #20232: Blazor FocusAsync
---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: aa724f22-3a49-c7f6-1500-ac2d8dde5894
* Version Independent ID: 5e8ef502-ccf9-659d-f25a-0f8f3c75f6e0
* Content: [ASP.NET Core Blazor event handling](https://docs.microsoft.com/en-us/aspnet/core/blazor/components/event-handling?view=aspnetcore-3.1)
* Content Source: [aspnetcore/blazor/components/event-handling.md](https://github.com/dotnet/AspNetCore.Docs/blob/master/aspnetcore/blazor/components/event-handling.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-blazor**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #20232
Fixes #20195

### CODE REVIEW

File: aspnetcore/blazor/call-javascript-from-dotnet.md [hunk 1]
```diff
@@ -192,7 +192,30 @@ The following example shows capturing a reference to the `username` `<input>` el
...
+```
+
+> [!NOTE]
+> Use [`FocusAsync`](xref:blazor/components/event-handling#focus-an-element) in C# code to focus an element, which is built-into the Blazor framework and works with element references.
+
+::: moniker-end
+
+::: moniker range="< aspnetcore-5.0"
+
+As far as .NET code is concerned, an <xref:Microsoft.AspNetCore.Components.ElementReference> is an opaque handle. The *only* thing you can do with an <xref:Microsoft.AspNetCore.Components.ElementReference> is pass it through to JavaScript code via JS interop. When you do so, the JavaScript-side code receives an `HTMLElement` instance, which it can use with normal DOM APIs.
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet.md:218]: Is this meant to still be here?

File: aspnetcore/blazor/call-javascript-from-dotnet.md [hunk 2]
```diff
@@ -192,7 +192,30 @@ The following example shows capturing a reference to the `username` `<input>` el
 >
 > If JS interop mutates the contents of element `MyList` and Blazor attempts to apply diffs to the element, the diffs won't match the DOM.
 
-As far as .NET code is concerned, an <xref:Microsoft.AspNetCore.Components.ElementReference> is an opaque handle. The *only* thing you can do with <xref:Microsoft.AspNetCore.Components.ElementReference> is pass it through to JavaScript code via JS interop. When you do so, the JavaScript-side code receives an `HTMLElement` instance, which it can use with normal DOM APIs.
+::: moniker range=">= aspnetcore-5.0"
+
+In most cases as far as .NET code is concerned, an <xref:Microsoft.AspNetCore.Components.ElementReference> is an opaque handle. For the most part, the only thing you can do with an <xref:Microsoft.AspNetCore.Components.ElementReference> is pass it through to JavaScript code via JS interop. When you do so, the JavaScript-side code receives an [`HTMLElement`](https://developer.mozilla.org/docs/Web/API/HTMLElement) instance, which it can use with normal DOM APIs.
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet.md:197]: Maybe it makes sense to just remove this "In most cases as far as .NET code is concerned" and just state how an ElementReference can be used when invoking code via JS interop.

**Comment by guardrex** [comment]
@captainsafia I was trying not to change Javier's lingo too much, but I agree with you. It's much better to just rewrite it to avoid the versioning altogether. Also, it was 😵 *half-baked* 😵 ... I didn't have **_all_** of the focus event-based example converted over to a click-based event example, which is necessary now that Blazor has a nice way to focus things without JS interop.

**Comment by guardrex** [comment]
Also @captainsafia ... one more little thing that we can put on here for issue #20195.

@mrlife inquired on the "guarantee" in the reasoning for not being able to pass element references between components. I think what that "guarantee" means that the element reference probably doesn't exist until the component is rendered ... therefore ... that late in rendering the component ... nothing can be passed as a parameter, including an element reference. Thus, the guidance is to use a callback.

I change that text here on the last commit to try and fix up the "guarantee" lingo to explain it that way. See if that update makes sense to u.

File: aspnetcore/blazor/call-javascript-from-dotnet.md [hunk 3]
```diff
@@ -192,40 +192,46 @@ The following example shows capturing a reference to the `username` `<input>` el
...
 
-As far as .NET code is concerned, an <xref:Microsoft.AspNetCore.Components.ElementReference> is an opaque handle. The *only* thing you can do with <xref:Microsoft.AspNetCore.Components.ElementReference> is pass it through to JavaScript code via JS interop. When you do so, the JavaScript-side code receives an `HTMLElement` instance, which it can use with normal DOM APIs.
-
-For example, the following code defines a .NET extension method that enables setting the focus on an element:
+An <xref:Microsoft.AspNetCore.Components.ElementReference> is passed through to JavaScript code via JS interop. The JavaScript code receives an `HTMLElement` instance, which it can use with normal DOM APIs. For example, the following code defines a .NET extension method that enables sending a mouse click to an element:
 
 `exampleJsInterop.js`:
 
 ```javascript
 window.exampleJsFunctions = {
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet.md:200]:
```suggestion
window.interopFunctions = {
```

File: aspnetcore/blazor/components/event-handling.md [hunk 4]
```diff
@@ -273,3 +273,27 @@ In the following example, selecting the check box prevents click events from the
...
+
+```razor
+<input @ref="exampleInput" />
+
+<button @onclick="ChangeFocus">Focus the Input Element</button>
+
+@code {
+    private ElementReference exampleInput;
+	
+    private void ChangeFocus()
```
Comment (captainsafia) [inline aspnetcore/blazor/components/event-handling.md:291]:
```suggestion
    private async Task ChangeFocus()
```

File: aspnetcore/blazor/components/event-handling.md [hunk 5]
```diff
@@ -273,3 +273,27 @@ In the following example, selecting the check box prevents click events from the
...
+    {
+        exampleInput.FocusAsync();
```
Comment (captainsafia) [inline aspnetcore/blazor/components/event-handling.md:293]:
```suggestion
        await exampleInput.FocusAsync();
```

File: aspnetcore/blazor/call-javascript-from-dotnet/samples_snapshot/component1.razor [hunk 6]
```diff
@@ -1,14 +1,17 @@
...
 @code {
-    private ElementReference username;
+    private ElementReference exampleButton;
 
-    public async Task SetFocus()
+    public async Task TriggerClick()
     {
         await JSRuntime.InvokeVoidAsync(
-            "exampleJsFunctions.focusElement", username);
+            "exampleJsFunctions.clickElement", exampleButton);
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet/samples_snapshot/component1.razor:15]:
```suggestion
            "interopFunctions.clickElement", exampleButton);
```

File: aspnetcore/blazor/call-javascript-from-dotnet/samples_snapshot/component1.razor [hunk 7]
```diff
@@ -1,14 +1,17 @@
 @inject IJSRuntime JSRuntime
 
-<input @ref="username" />
-<button @onclick="SetFocus">Set focus on username</button>
+<button @ref="exampleButton" />
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet/samples_snapshot/component1.razor:3]:
```suggestion
<button @ref="exampleButton" >Example Button</button>
```

File: aspnetcore/blazor/call-javascript-from-dotnet.md [hunk 8]
```diff
@@ -192,40 +192,46 @@ The following example shows capturing a reference to the `username` `<input>` el
...
 To use an extension method, create a static extension method that receives the <xref:Microsoft.JSInterop.IJSRuntime> instance:
 
 ```csharp
-public static async Task Focus(this ElementReference elementRef, IJSRuntime jsRuntime)
+public static async Task TriggerClickEvent(this ElementReference elementRef, 
+    IJSRuntime jsRuntime)
 {
     await jsRuntime.InvokeVoidAsync(
-        "exampleJsFunctions.focusElement", elementRef);
+        "exampleJsFunctions.clickElement", elementRef);
```
Comment (captainsafia) [inline aspnetcore/blazor/call-javascript-from-dotnet.md:225]:
```suggestion
        "interopFunctions.clickElement", elementRef);
```

**Review by captainsafia — COMMENTED** [review]
Looking pretty good!

Tangentially: do we have docs on our updated IJSObjectReference stuff?

**Comment by guardrex** [comment]
@captainsafia 

> do we have docs on our updated IJSObjectReference stuff?

Yes ... it's in the 5.0 version of the doc at ...

https://docs.microsoft.com/en-us/aspnet/core/blazor/call-javascript-from-dotnet?view=aspnetcore-5.0#blazor-javascript-isolation-and-object-references

... also mentioned here (and Pranav said this is enough here, i.e., we don't need an example for this mention) ...

https://docs.microsoft.com/en-us/aspnet/core/blazor/webassembly-performance-best-practices?view=aspnetcore-5.0#consider-making-synchronous-calls

**Comment by guardrex** [comment]
# 🙈😄

> `await exampleInput.FocusAsync();`

"FocusAsync" ....... **_Async!!_** .... `await` .... *Geebers McScreebers!* :smile:

One of these days, I'll learn how to do async. *Not today, of course!* ... but one day. 😁

**Comment by guardrex** [comment]
@captainsafia ... I committed ur updates. I think we're good here. This won't go live until later today/tonight, so we have time to sneak another PR in if needed.

**Pull request merged**